### DISCUSSION TITLE
A literal form for a list of name-value pairs

### DISCUSSION THREAD

**Discussion by ruv** [ForthHub/discussion#125] [General]
How would you **express** a list/map/set of key-value pairs (string, string) in Forth code, inside a definition?

In procedural style it could be expressed as:
```
    new-map ( mapid )
    dup "value 1" "name-1" -add-map-entry
    dup "value 2" "name-2" -add-map-entry
    dup "value 3" "name-3" -add-map-entry
    ( mapid )
```
Where `-add-map-entry ( mapid sd.value sd.name -- )` appends a pair into a map.

But I'm interested in a literal form that just puts this _mapid_ on the stack (as an example).

It doesn't matter how it can be implemented, only how it looks like in Forth code, and whether you like it.

As an example which I don't yet like much, this map can be expressed in a form of [S-expression](https://en.wikipedia.org/wiki/S-expression) as:
```
  (map
    (pair "name-1" "value 1")
    (pair "name-2" "value 2")
    (pair "name-3" "value 3")
  )
```
Or even slightly simpler:
```
  (map
    (name-1 "value 1")
    (name-2 "value 2")
    (name-3 "value 3")
  )
```

**Comment by SirWumpus** [comment]
Ignoring implementation and data-space concerns (can the map grow once new words are subsequently added), it would be nice to avoid the `DUP`s and simply have the word carry-over the `mapid`, eg. `map-pair  (mapid value key -- mapid)`:

```
map-new                        \ mapid
  " value1" " key1" map-pair   \ mapid
  " value2" " key2" map-pair   \ mapid
constant foo                   \ --
```

Would there be possible `map-tuple ( mapid valueN ... value0 n key -- mapid)`?

Maybe consider how the other operations on the `mapid` might look; maybe that might influence the literal form when seen as a whole.

**Reply by ruv** [reply]
>  it would be nice to avoid the `DUP`s and simply have the word carry-over the _mapid_, eg. `map-pair ( mapid value key -- mapid )`

Yes, and I even had exactly this variant in my draft.  But since it's not what I asked for, I left a straight but slightly awkward variant.
Usually, such a basic word has stack signature `( sd.value sd.key mapid -- )`.

Where _sd_ is a pair _( c-addr u )_ that represents a string.


> Would there be possible `map-tuple ( mapid valueN ... value0 n key -- mapid )`?

I have not grasped your idea. Do you mean to associate a key with a tuple (better say, with a list of single-cell values)?

> Maybe consider how the other operations on the mapid might look; maybe that might influence the literal form when seen as a whole.

In my use case it's an immutable map that is only used to fetch a value by a name (like a hash table). And in most cases of defining such a map there is no need to calculate names or values, they just are literals. Therefore the main question is how to represent the whole map in a literal form in code.

Actually, the choice lays between some well known serialization format (like JSON, S-expression, XML, etc) or a custom format.

If someone used a _literal_ form for such a task in Forth, what was your choice?

**Reply by SirWumpus** [reply]
> ... it's not what I asked for

Sorry.  I was unclear what was being asked; I assumed "how would you best express this".

> I  have not grasped your idea. Do you mean to associate a key with a tuple (better say, with a list of single-cell values)?

I was drawing on my Erlang experience concerning tuples and maps.  And yes, I was thinking of a tuple as a list of single-cell values.

> In my use case it's an immutable map that is only used to fetch a value by a name.

It sounds like a variant on `BEGIN-STRUCTURE` et al. which already allows fetching (and storing) a value by name (well, object base plus field offset).   How would a Forth map be much different from a structure (other than syntactic sugar)?

**Reply by ruv** [reply]
> How would a Forth map be much different from a structure (other than syntactic sugar)?

What I'm looking for is actually a variant of syntactic sugar 😅 

In a map, the number of pairs is uncertain; in my case the keys and values are arbitrary strings.
In a structure you have a fixed number of members, and the names of members are fixed symbols — these names cannot play role of keys.

**Comment by MitchBradley** [comment]
I use wordlists as maps.

vocabulary my-map
also my-map definitions
: key1 " value1" ;
: key2 " value2" ;
previous definitions

Then  
   ( adr len ) ['] my-map search-wordlist  if  execute  else ( whatever ) then

You can wrap it up in whatever syntactic sugar you want, or create the map entries programmatically, ...

**Reply by ruv** [reply]
> I use wordlists as maps.

Me too, sometimes.

> You can wrap it up in whatever syntactic sugar you want, or create the map entries programmatically, ...

Yes, except that this approach cannot be used during compilation (in a standard way; in a non standard — I use dynamically allocated dictionaries, so you can switch to such a dictionary, add anything into it, revert back and continue compilation of the definition).

But what I'm l looking for — is not implementations, but variants of **syntactic sugar**.

**Reply by SirWumpus** [reply]
```
oh-sweet-thing
  " jelly bean" " candy"  come-closer
  " roche" " chocolate" come-closer
constant ohh-lala
```
(sorry couldn't resist the jest)

**Comment by spelc** [comment]
```
[map
  "value 1" "value2"
  "value 3" "value 4" 
  ...
map]
```

Implementation details are left to the reader, but we (MPE/W&P) use a similar scheme for parsing C enums in Forth.

github appears to use a very peculiar version of Markdown.

**Reply by ruv** [reply]
Is it a list of only values, without names?

In what I'm looking for, the first compound entity is a pair of name and value, and the second — is a list of such pairs.

**Reply by spelc** [reply]
It's just a list of name/value pairs. The format can be as desired
[map

  name1   value1

name2  value2

...

map]

Since the parser is hidden in [MAP do what you like. Similarly, the terminator is hidden in MAP]. We often use 0 0 as a terminating pair.

**Reply by ruv** [reply]
@spelc, thank you — I see now.

> github appears to use a very peculiar version of Markdown.
 
 I have corrected formatting in your [message](https://github.com/ForthHub/discussion/discussions/125#discussioncomment-3832775) (see markup in the edit mode).

Markdown formatting can be tested at https://markdown-it.github.io/
